lib/validation_msg_deso_txn.go

```diff
@@ -0,0 +1,172 @@
+			"MaxUnconnectedTxSizeBytes %d", len(txnBytes), MaxUnconnectedTxSizeBytes)
+	}
+	return nil
+}
+
+func (m *MsgDeSoTxnValidator) ValidateTransactionMetadata() error {
+	// Validate that the transaction has correct metadata
+	if m.txn.TxnMeta == nil {
+		return fmt.Errorf("ValidateTransactionSanityBalanceModel: Transaction is missing TxnMeta")
+	}
+	if _, err := NewTxnMetadata(m.txn.TxnMeta.GetTxnType()); err != nil {
+		return errors.Wrapf(err, "ValidateTransactionSanityBalanceModel: Problem parsing TxnType")
+	}
+	return nil
+}
+
+func (m *MsgDeSoTxnValidator) ValidateHash() error {
+	// Validate transaction hash
+	if m.txn.Hash() == nil {
+		return fmt.Errorf("ValidateTransactionSanityBalanceModel: Problem computing tx hash")
+	}
+	return nil
+}
+
+func (m *MsgDeSoTxnValidator) ValidatePublicKey() error {
+	// Validate public key
+	if err := IsByteArrayValidPublicKey(m.txn.PublicKey); err != nil {
+		return errors.Wrapf(err, "ValidateTransactionSanityBalanceModel: Problem with public key")
+	}
+	return nil
+}
+
+func (m *MsgDeSoTxnValidator) ValidateFormatBalanceModel(blockHeight uint64) error {
+	// Validate transaction version
+	if m.txn.TxnVersion == DeSoTxnVersion0 {
+		return fmt.Errorf("ValidateTransactionSanityBalanceModel: DeSoTxnVersion0 is outdated in balance model")
+	}
+
+	if m.txn.TxnNonce == nil {
+		return errors.Wrapf(TxErrorNoNonceAfterBalanceModelBlockHeight, "ValidateFormatBalanceModel: Transaction "+
+			"does not have a nonce.")
+	}
+	if m.txn.TxnNonce.ExpirationBlockHeight < blockHeight {
+		return errors.Wrapf(TxErrorNonceExpired, "ValidateFormatBalanceModel: Transaction nonce has expired")
+	}
+	if m.globalParams.MaxNonceExpirationBlockHeightOffset != 0 &&
+		m.txn.TxnNonce.ExpirationBlockHeight > blockHeight+m.globalParams.MaxNonceExpirationBlockHeightOffset {
+		return errors.Wrapf(TxErrorNonceExpirationBlockHeightOffsetExceeded, "ValidateFormatBalanceModel: Transaction "+
+			"nonce expiration block height offset exceeded")
+	}
+
+	// Verify inputs/outputs.
+	if len(m.txn.TxInputs) != 0 {
+		return errors.Wrapf(RuleErrorBalanceModelDoesNotUseUTXOInputs, "ValidateTransactionSanityBalanceModel: Balance model "+
+			"transactions should not have any inputs")
+	}
+
+	// Loop through the outputs and do a few sanity checks.
+	var totalOutNanos uint64
+	for _, txout := range m.txn.TxOutputs {
+		// Check that each output's amount is not bigger than the max as a
+		// sanity check.
+		if txout.AmountNanos > MaxNanos {
+			return errors.Wrapf(RuleErrorOutputExceedsMax, "ValidateTransactionSanityBalanceModel: Output amount %d "+
+				"exceeds max %d", txout.AmountNanos, MaxNanos)
+		}
+		// Check that this output doesn't overflow the total as a sanity
+		// check. This is frankly impossible since our maximum limit is
+		// not close to the max size of a uint64 but check it nevertheless.
+		if totalOutNanos >= math.MaxUint64-txout.AmountNanos {
```

Comment:
This function never increases totalOutNanos. You can use the `SafeUint64` functions to do a safe add that throws an error if it overflows.

Reply:
Nice! Forgot `SafeUint64` existed. I copied this code from the BlockView connect logic. One difference is that the BlockView will error if `totalOutNanos + txout.AmountNanos == math.MaxUint64` while `SafeUint64` won't. Don't think it matters tho.